### DISCUSSION TITLE
Controlling Re-rendering behavior

### DISCUSSION THREAD

**Discussion by matthewcrews** [fsprojects/Avalonia.FuncUI#341] [Q&A] [answered]
I have an Elmish app split into two areas: Diagram and Charting. I need the diagram area to refresh as I move the mouse around continually, but I want to keep the Charting area from continually refreshing. Is there some way to indicate that a particular portion of an Elmish UI doesn't need to be re-rendered? I tried using the `View.createWithKey` to see if that would help, but it did not.

![image](https://github.com/fsprojects/Avalonia.FuncUI/assets/1593215/a6fb9d48-376a-4ab3-b3af-7c544fe3c1c8)

**Comment by JaggerJo** [comment, marked as answer]
Hey @matthewcrews,

there are multiple ways of doing this. I assume you also want to only diff certain parts of the view tree.

There is a setting in the Renderer you can enable for debugging. That way you see which parts are really redrawn.
```fsharp
type MainWindow() as this =
    inherit HostWindow()
    do
        ...
        this.RendererDiagnostics.DebugOverlays <-
            Rendering.RendererDebugOverlays.Fps |||
            Rendering.RendererDebugOverlays.DirtyRects
```

The easiest solution to your problem is using the component model. All you need to do is wrap the chart views in a component that takes the charting data as an argument.

```fsharp
let chartView (data: ChartData) =
    Component.create ($"key-{data.Id}",, fun ctx -> 
        Chart.create [
            ...
        ]
    )
    
let mainView (model, dispatch) =
    StackPanel.create [
        StackPanel.children [
            ... node view ...

            chartView model.SelectedNode.ChartData
        ]
    ]

```

**Reply by matthewcrews** [reply]
Beautiful! The only thing I'll note for those following this is that I had to wrap the `Component` in another layout object to specify the `Grid` positioning. Originally I had the grid positioning in the `Chart` component. `Grid.row` and `Grid.column` needed to exist outside the Component for the layout to work.

**Reply by JaggerJo** [reply]
You can also set `Grid.row/Grid.column` in the component ([Docs](https://funcui.avaloniaui.net/common-questions#how-to-set-attributes-on-component-level)).

```fsharp

Component(fun ctx ->
    ctx.attrs [
        Component.row 1
        Component.column 1
    ]
    Button.create []
)
```

**Reply by matthewcrews** [reply]
I am confirming this works as well. It looks like the following:

```fsharp
Component.create ($"key-{nodeName}", fun ctx ->
            ctx.attrs [
                Grid.row 1
                Grid.column 0
            ]
            Chart.create ...
)
```
